enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,612 @@
+hugepages1G = 0
+
+# Additional kernel arguments
+additionalArgs = ""
+```
+
+#### CRI-O configuration
+
+```ini
+[crio.runtime.runtimes.high-performance]
+runtime_path = "/bin/crun"
+runtime_type = "oci"
+runtime_root = "/bin/crun"
+allowed_annotations = ["cpu-load-balancing.crio.io", "cpu-quota.crio.io", "irq-load-balancing.crio.io", "cpu-c-states.crio.io", "cpu-freq-governor.crio.io"]
+```
+
+#### Kubelet configuration 
+
+Because of multitude of option in Kubelet configuration, a simple passthrough (copy paste)
+will be implemented, rather than exposing every single little configuration variable.
+
+```yaml
+# /etc/microshift/config.yaml
+kubelet:
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"
+  cpuManagerReconcilePeriod: 5s
+  memoryManagerPolicy: Static
+  topologyManagerPolicy: single-numa-node
+  reservedSystemCPUs: 0,28-31
+  reservedMemory:
+  - limits:
+      memory: 1100Mi
+    numaNode: 0
+  kubeReserved:
+    memory: 500Mi
+  systemReserved:
+    memory: 500Mi
+  evictionHard:
+    imagefs.available: 15%
+    memory.available: 100Mi
+    nodefs.available: 10%
+    nodefs.inodesFree: 5%
+  evictionPressureTransitionPeriod: 0s
+```
+will be passed-through to kubelet config as:
+```yaml
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  full-pcpus-only: "true"
+cpuManagerReconcilePeriod: 5s
+memoryManagerPolicy: Static
+topologyManagerPolicy: single-numa-node
+reservedSystemCPUs: 0,28-31
+reservedMemory:
+- limits:
+    memory: 1100Mi
+  numaNode: 0
+kubeReserved:
+  memory: 500Mi
+systemReserved:
+  memory: 500Mi
+evictionHard:
+  imagefs.available: 15%
+  memory.available: 100Mi
+  nodefs.available: 10%
+  nodefs.inodesFree: 5%
+evictionPressureTransitionPeriod: 0s
+```
+
+#### Extra manifests
+
+Connects Pod's `.spec.runtimeClassName` to CRI-O's runtime.
+If Pod has `.spec.runtimeClassName: microshift-low-latency`,
+it can use annotations specified in CRI-O config with `crio.runtime.runtimes.high-performance`.
+
+```yaml
+apiVersion: node.k8s.io/v1
+handler: high-performance
+kind: RuntimeClass
+metadata:
+  name: microshift-low-latency
+```
+
+
+### Risks and Mitigations
+
+Biggest risk is system misconfiguration.
+It is not known to author of the enhancement if there are configurations (like kernel params, sysctl, etc.)
+that could brick the device, though it seems rather unlikely.
+Even if kernel panic occurs after staging a deployment with new configuration,
+thanks to greenboot functionality within the grub itself, the system will eventually rollback to
+previous deployment.
+Also, it is assumed that users are not pushing new image to production devices without prior verification on reference device.
+
+It may happen that some users need to use TuneD plugins that are not handled by the profile we'll create.
+In such case we may investigate if it's something generic enough to include, or we can instruct them
+to create new profile that would include `microshift-low-latency` profile.
+
```

Comment:
Regarding day2 operations and updates: we need to make sure that if we provide any updates (e.g. to profiles), we do not overwrite stuff that the customer has modified. There has to be  a clear seperation of responsibilities what we deliver and what the customer uses. We deliver a good baseline for the customer to start their tuning. If we change that, we should not touch customer tuning. E.g. by layering the config options, and applying customer values last.  I think we have this, but I would feel better if it would be stated explicitliy somewhere. (Probably not here, just came to my mind)

Reply:
There is already something like this to protect user edited config files in the RPM specification: https://www.cl.cam.ac.uk/~jw35/docs/rpm_config.html
This is when the edit "ours" `/etc/tuned/microshift-low-latency-variables.conf`. If they supply their own profile with `/etc/tuned/my-profile-variables.conf` - we won't even know (except for the setting to autoapply the profile)